### ISSUE TITLE
GEMM fails to run on a multi-GPU system

### ISSUE THREAD

**Issue by kpot** [CNugteren/CLBlast#155]
Hello Cedric,

I ran into a very weird behavior of CLBlast on my machine with couple of Radeon HD 7970 (Tahiti). When I run GEMM using only one GPU - it works flawlessly. But when I try to run GEMM sequentially on two GPUs - it fails on the second one, printing out `CLBlast: OpenCL error: clGetKernelWorkGroupInfo: -48`. I spent some time trying to rule out possible issues with my code or system, but found none. BTW, I use Ubuntu 15.10 with the latest available version of radeon-crimson (fglrx) driver 15.12 and had no problems with OpenCL so far.

Here's my test I wrote to isolate and reproduce the bug. The code should be pretty straightforward.
Please, find some time to check it.
Again, GEMM fails only when I use both GPUs, in any order. I added run of a simple kernel into the test just to be sure that the bug isn't in OpenCL itself. It isn't.
I haven't checked any other CLBlast functions though, perhaps the bug would appear with them as well.
UPD: tested on both master and 0.11.0 branches.
```
#include <algorithm>
#include <iostream>
#include <string>
#include <vector>

#include "CL/cl.hpp"
#include <clblast.h>

// Sizes of the matrices
constexpr int m = 4, n = 5, k = 6;

// Controls which devices we're going to use
const int try_on_devices[] = {0, 1};

int main() {
  // Initialization
  std::vector<cl::Platform> platforms;
  std::vector<cl::Device> devices;
  cl::Platform::get(&platforms);
  platforms[0].getDevices(CL_DEVICE_TYPE_GPU, &devices);
  cl::Context context(devices);
  std::cout << "Detected devices: " << std::endl;
  for (auto &device : devices) {
    std::string device_name;
    device.getInfo(CL_DEVICE_NAME, &device_name);
    std::cout << device_name << std::endl;
  }
  std::cout << std::endl;

  // Running some test kernel to check if we can use both cards at all
  // (to check OpenCL driver itself)
  std::cout << "Running test kernel... " << std::endl;
  std::string source = "__kernel void the_answer(__global int *v) { "
                       "v[get_global_id(0)] = 42; }";
  auto program = cl::Program(context, source, true);
  auto the_answer = cl::make_kernel<cl::Buffer &>(program, "the_answer");
  for (int device_index : try_on_devices) {
    cl::CommandQueue queue(context, devices[device_index]);
    std::vector<int> host_buffer(64, 0);
    cl::Buffer test_buffer(context, CL_MEM_READ_WRITE,
                           sizeof(cl_int) * host_buffer.size());
    auto execution_is_done =
        the_answer(cl::EnqueueArgs(queue, cl::NDRange(host_buffer.size()),
                                   cl::NDRange(host_buffer.size())),
                   test_buffer);
    execution_is_done.wait();
    queue.enqueueReadBuffer(test_buffer, CL_TRUE, 0,
                            sizeof(cl_int) * host_buffer.size(),
                            host_buffer.data());
    bool it_works = std::all_of(host_buffer.begin(), host_buffer.end(),
                                [](int x) { return x == 42; });
    std::cout << "Device " << device_index << ": simple kernel "
              << (it_works ? "works well" : "doesn't work") << std::endl;
  }

  // Trying to do the same with GEMM
  std::cout << std::endl << "Running GEMM now" << std::endl;
  for (int device_index : try_on_devices) {
    cl::CommandQueue queue(context, devices[device_index]);
    std::vector<float> mat_A(m * k);
    std::vector<float> mat_B(k * n);
    std::vector<float> mat_C(m * n);

    for (auto &item : mat_A) { item = 1.0f; }
    for (auto &item : mat_B) { item = 1.0f; }
    for (auto &item : mat_C) { item = 0.0f; }

    cl::Buffer mat_A_buffer(context, CL_MEM_READ_WRITE,
                            sizeof(float) * mat_A.size());
    cl::Buffer mat_B_buffer(context, CL_MEM_READ_WRITE,
                            sizeof(float) * mat_B.size());
    cl::Buffer mat_C_buffer(context, CL_MEM_READ_WRITE,
                            sizeof(float) * mat_C.size());

    queue.enqueueWriteBuffer(mat_A_buffer, CL_TRUE, 0,
                             sizeof(float) * mat_A.size(), mat_A.data());
    queue.enqueueWriteBuffer(mat_B_buffer, CL_TRUE, 0,
                             sizeof(float) * mat_B.size(), mat_B.data());
    queue.enqueueWriteBuffer(mat_C_buffer, CL_TRUE, 0,
                             sizeof(float) * mat_C.size(), mat_C.data());

    cl_command_queue cl_queue = queue();
    cl_event work_is_done = cl_event{nullptr};

    std::cout << "Multiplying using device " << device_index << std::endl
              << "--------------------------" << std::endl;
    auto status =
        clblast::Gemm(clblast::Layout::kColMajor, clblast::Transpose::kNo,
                      clblast::Transpose::kNo, m, n, k, 1.0f, mat_A_buffer(), 0,
                      m, mat_B_buffer(), 0, k, 0.0f, mat_C_buffer(), 0, m,
                      &cl_queue, &work_is_done);
    if (status == clblast::StatusCode::kSuccess) {
      clWaitForEvents(1, &work_is_done);
      clReleaseEvent(work_is_done);
      queue.enqueueReadBuffer(mat_C_buffer, CL_TRUE, 0,
                              sizeof(float) * mat_C.size(), mat_C.data());
    } else {
      std::cerr << "Cannot multiply matrices using device "
                << std::to_string(device_index);
      return 1;
    }

    std::cout << "Result: (all elements must be equal to " << k << ")"
              << std::endl;
    for (int i = 0; i < m; ++i) {
      for (int j = 0; j < n; ++j) {
        std::cout << mat_C[j * m + i] << " ";
      }
      std::cout << std::endl;
    }
    std::cout << std::endl;
  }

  return 0;
}
```

**Comment by CNugteren** [comment]
Hi there. Thanks for the detailed message!

Previously another user reported a similar issue in #133, but since there was no more response and I couldn't reproduce it, I closed it. Nevertheless, I still have the 'test' I produced back then for myself. I've updated that branch (`multithreading_test`) to contain the latest changes from master as well. Perhaps you can check out the branch and try the [test program](https://github.com/CNugteren/CLBlast/blob/multithreading_test/test/correctness/misc/multithreading.cpp): it is automatically compiled when compiling the tests and is named `clblast_test_multithreading`. However, it actually failed on my multi-platform system just now...

I will look at your issue in a bit more detail and try out your code in a few days on a multi-GPU system. I'm afraid I don't have access to a multi-AMD-GPU unfortunately.

**Comment by kpot** [comment]
Hi Cedric, thanks for the quick reply!

I've checked `multithreading_test` branch as you asked and it doesn't work as well. Frankly, I didn't hope that would help anyway, since the problem is clearly not related to threads. I run GEMMs sequentially, each time waiting the event before starting the next iteration on another GPU.

But during my experiments I found a clue: calling `clblast::ClearCache()` after each GEMM resolves the problem (I checked on both `multithreading_test` and `master` branches). Apparently it has to do something with the cache. Also, if I correctly decipher the message `clGetKernelWorkGroupInfo: -48`, it means `CL_INVALID_KERNEL`. I believe caching somehow fails to store the same kernel for both devices. Or, since the devices are identical, probably the `Program` object gets corrupted during the second run of GEMM, when it checks if the program is already cached. Of course, calling `clblast::ClearCache()` all the time is too terrible, so I hope you have an idea of what might be the issue.

**Comment by CNugteren** [comment]
OK, thanks for investigating! There are two levels of caching in the library, and I believe one might be device specific. Perhaps that doesn't work well with two devices with the same name? I'll check it out as soon as possible. You can expect an update from my side no later than Wednesday evening.

**Comment by kpot** [comment]
Hi,
I've investigated it a little bit more. Particularly, [I've tried to replace clblast::Device::Name()](https://github.com/kpot/CLBlast/commit/e9b3981d078879555feaa1bec7cb5b08faac9f91) to make sure each device has its unique name connected with its location on PCIe bus. The names look slightly ridiculous now (like `Tahiti 010000000000000000000000000000000000000000010000` or `GeForce GTX 660 Ti 0300000000000000`) but they're stable and unique, and I ensured that `BinaryCache::Instance().Get(...)` call uses them.
Unfortunately, it didn't help at all. I still get the same `OpenCL error: clGetKernelWorkGroupInfo: -48`.

**Comment by CNugteren** [comment]
So there are two levels of caches in CLBlast:

1. A binary cache based on [the device name](https://github.com/CNugteren/CLBlast/blob/master/src/cache.hpp#L70).
2. A program cache based on [the OpenCL context](https://github.com/CNugteren/CLBlast/blob/master/src/cache.hpp#L82).

There might be two issues with this and your multi-GPU test set-up. First, the binary cache might indeed need a unique identifier (e.g. OpenCL device pointer). Second, the program cache might get into problems when the OpenCL context contains multiple devices.

I tested on a dual NVIDIA system with twice the same device name, and your test didn't work. Then I changed the code such that it creates a new context for each device (instead of one context for all devices) and then it works. I've added this code to your sample program at the beginning of the body of the `device_index` loop:
```
    cl::Device this_device = devices[device_index];
    cl::Context context({this_device});
```
This 'fixes' the second issue I mentioned above. Apparently the first 'issue' is not an actual issue on my machine, not sure if it would be on yours.

Could you test if you can get it working using the above code in your sample? If so, I'll make a fix in the code to properly handle contexts with multiple devices. And if you can also verify whether the first issue is an actual issue (e.g. by using the PCI Express ID as you did) then I'll also solve that. Thanks.

**Comment by kpot** [comment]
Hi Cedric,
I think I have a solution.

Creating multiple contexts isn't the best way, since it makes impossible to implement many useful optimizations. Having one context for multiple devices is a powerful thing. But I remembered that `clBuildProgram` actually accepts multiple devices as one of its arguments and the spec says that 

> "device_list: A pointer to a list of devices that **are in program**".

So the device is an inseparable part of the program. If you need to run the same kernel on multiple devices, you need to either build multiple programs for each of them, or build the same program once but for all your devices together.

And here lies the problem. CLBlast currently builds each program only once for a given context and precision. And only for a single device (the first one), because it detects the device from the queue, which can be associated with only one of them. At the moment of building there's no ways to know which other device the user is going to use next. As result, when we try to perform the same program/kernel on an another device, we can't do that because the device *is not in the program* we built previously.

So I decided to change this part and added the device_name into the program cache`s key. Here I mean my previously extended device_name with the "topology part" added to it, to make it unique (otherwise it doesn't make much sense). And now it works! You can find my changes [here](https://github.com/kpot/CLBlast/commit/b7c14d88532702849799f76bd6f5f5f7d8a8dad1), they're very simple.

So, to recap all above: we need two things to make it work:

1. Unique device names. I used combination of `CL_DEVICE_NAME` and PCIe topology data provided by OpenCL drivers (and tested on NVidia and AMD drivers). I'm not sure the `cl_device_id` pointer can work as an unique identifier and stay consistent between `clGetDeviceIDs` calls.
2. Build and cache program for each device individually.

What do you think?

**Comment by CNugteren** [comment]
Yes, it is clear that there is something wrong. But it isn't clear yet what a good solution would be. Indeed, if you add the device name to the program-cache, you solve the issue with a context with multiple devices. But then what's now the use of the binary-cache? If the binary-cache checks on a subset of IDs (device name, routine name, precision) used for the program-cache, then we could actually remove one of the caches (binary-cache is then preferred of course). But yeah if this is really needed then we'll have to remove one of the caches.

The other issue I have is with the device name + PCI Express ID. First of all that doesn't work on all platforms so it's not a nice solution in the end. Also, I think we would need to add platform name as well since a device with a specific name and PCI Express ID might exist on multiple platforms. But most importantly, don't you think this should actually work well without the PCI Express ID? Can't a single binary be run on two identical devices? And if there is indeed something funky in the binary related to the device, then surely a program must be able to run on multiple instances of a specific devices?

**Comment by kpot** [comment]
Hi, I have what I believe is this better solution, which addresses your concerns. It actually already works but I want to test it bit more. On Thursday I'm going to finish it and commit the result with detailed explanation.

**Comment by kpot** [comment]
Your last comment encouraged me to dig deeper in how all this machinery works, and I focused on three questions:

1. Can I reproduce the same bug we see in CLBlast in my own code?
2. Can I re-use the same binary with multiple devices?
3. What else I can use to identify the device instead of the device name and PCIe location? So I could build my own program/binary cache.

**A1**. The program is really associated with devices, so my previous statements about it are correct. It seems that "under the hood" `clBuildProgram` not only builds the program, but actually *delivers* it on the devices, similar to what `clEnqueueWriteBuffer` does.
Even when we create the program instance from an already pre-compiled binary using `clCreateProgramWithBinary`, we still must then call `clBuildProgram` on this program. And calling either of these methods we must pass to it the device we're going to run this program on. Otherwise it won't work, even if this binary was previously built for the very same device. It fails exactly like CLBlast does.

**A2**. Yes, for as long as you call `clBuildProgram` after `clCreateProgramWithBinary`, passing the device you're going to use, you will get a program that you can run on this device. But this program won't work on an another device of the same type as the first one, unless you specify this device too in the arguments. All devices having the same name can re-use/share binaries. In theory, some different vendors can have identically named devices. But it seems so improbable, and even less improbable to see such devices on a single machine, that I decided not to bother adding vendor`s name to the cache.

**A3**. Surprisingly, `cl_device_id` is a stable identifier within the same program run, and the same is true for `cl_context` during its lifetime. I tested it with NVidia, AMD and Intel OpenCL drivers on Ubuntu and Mac. Technically, both types are pointers to some internal driver's structures. And they remain the same for the same devices no matter what. It seems logical and obvious, otherwise the memory management would become a hell for the developers of the driver, but I decided to test it anyway. I tried to list platforms and devices hundreds of times over large periods of time. Also tried to fetch CL_QUEUE_DEVICE and CL_QUEUE_CONTEXT from newly created queues. These pointers always contain the same memory addresses during lifetime of the process (or OpenCL context for `cl_context`). This means we can safely use both cl_device_id and context_id as map`s keys to cache the programs. And even if we assume that the identifiers may change sometimes, for as long as they remain the same for the queue (and I believe they will), it won't affect CLBlast's performance.

Finally, I wrote a "proof of concept" in C++ which implements both binary and program caching for all GPUs of all platforms available in the system. So you can also play with it. And I [modified CLBlast accordingly](https://github.com/kpot/CLBlast/commit/62f03ffd33d026bf58a194f93458ac34e0076f96), it also works well. Here's my PoC:

```
#include <algorithm>
#include <iostream>
#include <map>
#include <string>
#include <tuple>
#include <vector>

#include "CL/cl.hpp"

// Controls how many times we're going to run the kernel for all queues/devices
const int cycles_over_all_queues = 5;

using BinaryCache = std::map<std::string, std::vector<char>>;
using ProgramCacheKey = std::tuple<cl_context, cl_device_id, std::string>;
using ProgramCache = std::map<ProgramCacheKey, cl::Program>;

class CodeCache {
public:
  cl::Program get_program(const cl::Context &context, const cl::Device &device,
                          const std::string &program_name,
                          const std::string &source) {
    cl::Program program;
    ProgramCacheKey program_cache_key =
        std::make_tuple(context(), device(), program_name);
    auto search = program_cache_.find(program_cache_key);
    if (search != program_cache_.end()) {
      program = search->second;
      std::cout << "Using program from cache." << std::endl;
    } else {
      std::cout << "No matching programs in cache. Building one." << std::endl;
      std::string device_name;
      device.getInfo(CL_DEVICE_NAME, &device_name);
      auto search = binary_cache_.find(device_name);
      if (search == binary_cache_.end()) {
        std::cout
            << "No cached binary for the device. Building one from source code."
            << std::endl;
        program = cl::Program(context, source, false);
        program.build({device});
        size_t binary_size;
        clGetProgramInfo(program(), CL_PROGRAM_BINARY_SIZES,
                         sizeof(binary_size), &binary_size, nullptr);
        std::vector<char> binary_data(binary_size);
        char *program_pointers[1] = {binary_data.data()};
        clGetProgramInfo(program(), CL_PROGRAM_BINARIES,
                         sizeof(program_pointers), &program_pointers, nullptr);
        binary_cache_[device_name] = std::move(binary_data);
      } else {
        std::cout << "There's already cached binary for the device. Using it "
                     "to build the program."
                  << std::endl;
        auto &binary = binary_cache_[device_name];
        size_t binary_size = binary.size();
        cl_device_id cl_device = device();
        const unsigned char *binary_ptr = (unsigned char *)binary.data();
        program = clCreateProgramWithBinary(
            context(), 1, &cl_device, &binary_size, &binary_ptr, nullptr, nullptr);
        program.build();
      }
      std::cout
          << "Newly created program has been cached for this device and context"
          << std::endl;
      program_cache_[program_cache_key] = program;
    }
    return program;
  }

private:
  // Stores all binaries in a map using CL_DEVICE_NAME`s value as a key
  BinaryCache binary_cache_;
  // Stores all programs in a map using both cl_context and cl_device_id as keys
  ProgramCache program_cache_;
};

int main() {
  // Initialization of all queues for all GPUs of all platforms available
  std::vector<cl::Platform> platforms;
  std::vector<cl::Device> all_devices;
  std::vector<cl::Context> all_contexts;
  std::vector<cl::CommandQueue> all_queues;
  cl::Platform::get(&platforms);
  for (auto &platform : platforms) {
    std::vector<cl::Device> devices;
    platform.getDevices(CL_DEVICE_TYPE_GPU, &devices);
    if (devices.size() > 0) {
      std::string platform_name;
      platform.getInfo(CL_PLATFORM_NAME, &platform_name);
      std::cout << "Devices detected on platform " << platform_name << ":"
                << std::endl;
      cl::Context context(devices);
      for (auto &device : devices) {
        std::string device_name;
        device.getInfo(CL_DEVICE_NAME, &device_name);
        std::cout << device_name << " (available via queue "
                  << all_queues.size() << ")" << std::endl;
        all_queues.emplace_back(context, device);
      }
    }
  }
  std::cout << std::endl;

  // A test kernel
  std::string source = "__kernel void the_answer(__global int *v) { "
                       "v[get_global_id(0)] = 42; }";

  CodeCache cache;
  for (int main_loop = 0; main_loop < cycles_over_all_queues; ++main_loop) {
    for (int queue_index = 0; queue_index < all_queues.size(); ++queue_index) {
      std::cout << "Running test kernel on queue " << queue_index << std::endl
                << "--------------------------------" << std::endl;
      auto queue = all_queues[queue_index];
      cl::Context context; queue.getInfo(CL_QUEUE_CONTEXT, &context);
      cl::Device device; queue.getInfo(CL_QUEUE_DEVICE, &device);
      auto program = cache.get_program(context, device, "my_program", source);
      auto the_answer = cl::make_kernel<cl::Buffer &>(program, "the_answer");
      std::vector<int> host_buffer(64, 0);
      std::cout << "Host buffer size: " << host_buffer.size() << std::endl;
      cl::Buffer test_buffer(context, CL_MEM_READ_WRITE,
                             sizeof(cl_int) * host_buffer.size());
      auto execution_is_done =
          the_answer(cl::EnqueueArgs(queue, cl::NDRange(host_buffer.size()),
                                     cl::NDRange(host_buffer.size())),
                     test_buffer);
      execution_is_done.wait();
      queue.enqueueReadBuffer(test_buffer, CL_TRUE, 0,
                              sizeof(cl_int) * host_buffer.size(),
                              host_buffer.data());
      bool it_works = std::all_of(host_buffer.begin(), host_buffer.end(),
                                  [](int x) { return x == 42; });
      std::cout << "Device " << queue_index << ": simple kernel "
                << (it_works ? "works well" : "doesn't work") << std::endl
                << std::endl;
    }
  }
  return 0;
}
```

**Comment by CNugteren** [comment]
Great! Thanks a lot for your thorough research! I'll take a look at your changes and at you sample code when I have the time, hopefully tomorrow evening. In the meantime feel free to make it a pull request if you think that's a good idea.

**Comment by CNugteren** [comment]
This is now fixed thanks to PR #157.
